### DISCUSSION TITLE
How to properly update the error type of a query/mutation result?

### DISCUSSION THREAD

**Discussion by Ionshard** [TanStack/query#5784] [Q&A] [answered]
I am getting started with react-query and we have an Axios API library automatically generated from Swagger.

Additionally many many of our REST endpoints always return the same data structure on an error and we have our own internal error codes. To handle the myriad of errors that can possibly go wrong when making an API call, I have created a `parseError` utility function that takes an `unknown` error and will return a discriminated union that categorizes the error into many possible errors (one of which is our standard API error response). This utility function is extremely useful when combined with `ts-pattern` which allows us to write very clean and exhaustive error code.

Here is a sort of mock example of what this gives us where we can super easily match our standard API error for an INVALID_FILTER code and extract the detailedMessage from the data for use. Or we match a 403 response, otherwise we handle the unknown error.
```tsx
const TaskInfo = ({id}: {id: number}) => {
  const { data, isLoading, isError, error } = useQuery({
    queryKey: ['getTask', id],
    queryFn: () => getTask(id)
  });

 if(isLoading) return <Loader />

 if(isError) return match(parseError(error))
    .with({type: 'ApiError, errorCode: INVALID_FILTER, detailedMessage: P.select()}, (message) => <InvalidFilterError message={message} />)
    .with({type: 'HttpError', code: 403}, () => <NotAuthorisedError />)
    .otherwise(() => <UnknownError />)

  return <TaskContent task={data} />
}
```

Now my problem. `useQuery` by default has an unknown error type. When creating our custom hooks that use react-query I would love to handle the automatic error parsing as part of it, since we are a growing team making sure everyone remembers to parse the error and handle all potential issues can be difficult and hopefully I can eventually just auto generate these hooks from the swagger as well.

However, I can not for the life of me figure out how to properly type a hook that updates the error type.

Essentially what I am looking at is how I could write a hook like this, but without Typescript errors.

```tsx
const useGetTask = (id: number): UseQueryResult<Task, ErrorResult> => {
  const result = useQuery({
    queryKey: ['getTask', id],
    queryFn: () => getTask(id)
  });

  if(result.isError){
    result.error = parseError(result.error);
    return result;
  }

  return result;
}
```

Here is a [reproduction](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVQM4FMCKBXdUCeASuqtgDYwA0c2GOe+cAvnAGZQQhwBEAAjAEMAdqkEBjANYB6KOgFiYAWgCOuAtwCwAKG2hIsRHACCAD2ARUAUSgco1YKlPmrN6MzYcuAcgFmLXgG5dcGh4JAdLcBhGFnZOOC8yCAATAVQAC0Dg-TC4AC0U61s4NPzC1ygAFXwwdGoAL3c473qUrJ0tKSk4AF4+-oHBoYHtLrgwASgMIrdsGGAyYGjR7uG19Z7tdBMcuDEIEXgZqGJUSBF0ABEBQQBlMXT0EAFeuHrtODgAOggAIwArdAKAAUCA+nzgeFsAElkgAuN5fURQYBCADmwIAlD8wPMDgIyFjKODPlDoABhFLoBH1L5CbAgX54LE4vFCAlEkmQioAWRIqAEaOpiORqIx2IguPM7MJmOJWghcGS6EEC3QyT5qAFQppSJgKPRLMlbI5cq5ADcCcBUmzjqhddghBIhBAAO5CFmTKACfBGqX42Xyz5MTHgvUokGhhXfWSsVHoYHAm0CTG9AB8cAAhBEor7k3LDCB+YLhV5jqdzhhroI9sIXfAmZDc15mJigh1trtorU4OWSJWrjcXj1EajWHgADzd9AQVi9ioVg5Vof3R7PNPtrY7UJ7JfwCIL-tLwc1kdJocIx3Ot1CTEI5NwBzz2yLi7V4dp8F9s7H9+rp4CEiAjjgACpMGDnoI2KkGIYj8puWidju+yHM+0Cvhg-7PK8tJ-ICIJgtGog3LQur0oyzJmtGyYIt+A5-g8AHyiGCFIQY05oScR4XDhXxjpO06zpxGHoFhAgbtoW67ChoiPi4L7cRgrzArIP4XJeTouu6d5wKpA5ycJinoOmX6HmpmGMc8QGgeBCZ6ce0HYLB8GSR0UgAFTuR87nGHAxFCKkUDJMA9TqtyxTNHAMCPHAABSRjGGAwBwLceDmng3lSFJO4cUYSXHK8hGfNOCJeHlwDHO0nw0YZ5kngI7akhUAD8CJOBYxztkwCEeV5Co+QlAASlSVCBnEePE7WoJl2XsTUxmDTAMBgAVI5TQVABkiDgiVCSLctlVdT1nneb5sUCJaqBiCiuKcTNiHbnNPbnZd13ALiq3bdGu1eC9aRvR9FRVeF0C0RUR2ub1p2VDFYg3A8igEmQIN2HA6SSugrDkGQjAZBA5DJHAQjoOlUBwI2GDoEIXz3Wx8Accgmk3p9RVRfNpWM9e7qHeCZJQBpXNCBDHaPfT821eQ8BrflFRwAAPnA+0rbLCt-VdN0wAVCuc1pQida5Mn7qgasA5rstnnzAu6zpfMGatGa26i-lwUJ+tuSd-VwAAYo6CjSlFEDjLZJRCDQTPuijUXpDccAQLB2CyITyQJ2KIfGCB0K6egqgkPuQgwIHAincF6ugKiNxhY6-tCZAWrAL8ZDGXzqA0-1p3Qkbfn6gcaI41nsGcEWAVhQXQfAGT0VPljQh+wccCuks6RwAABs8MAPMvE0gKdy8wKgigTEteBCJvrBuNs0e0PM6WR9HAWLOirdwH1z9wLw2wCOAjc78vu-zaXuJTp6B3EgNeDxqBjViJ4BIe8D43BgMfKqPkuhfFQadfUjAio+U+GMHkAgJDGSMBnOAxAc6iFOiwOG68l7Aj5lGV+u5UJ801NqYyI4wHpGBBMKY6Bji0IqJieh2CvgL2isCU6EIkA-XKpVagNUkDMOLDqOAIEkToEbiCVRop0SphYEwIM2DPj8NsCwksqYegO15EouoEihGfB+NFPAC8IJYnTAkW4nAVTpFTq6Km8BXQcEfl4Ns3kJEFzSDAL4fNjHQFMUKEJnsmA-2XrTUWjDZLcOmObOAMT+Zh0Fjpb8ks3Gs2AHOYEB4FJ1VyYIrOMAE6hykezBIMigbUFQVE2W3VXKfDKTkyp6EjLvhqamWQ9SoCNLZrUUqrTbBeDkReSO3SOi9PKQ4daFQRlfUVIbLOdVXh8y+PZC4CFFR9IqfJQZ1TjkYFqWMhphhpEyzmdQPm7TUE3OMssrk9yJmPOaV4JWsjvioNtss4MPTHxrONhdf6Gs+F0NGSqB5TTpkJBNvCtpSzTl1JRVyH6OtmZtK5G88EyzvmdFWBsal-QVifFYPgwhGc6VwBhk+fYyoDJIzgEKYm3oEGEwEHMTgNxgBwzIDjTMdKaUytchxSoaQJCFXBNaBE5EmR2HBOyIsCJtFoiDEqEgptpS6v1GKA1GMhAIl+BACAjdhDaCSR0XZQoYAKtQEqs8qqiYMg1TpECngHDoAnO6iQGYLEqMDRgI5JA7XpVBNaag2rSxe0ZayxV8zDUALZDMtgabBAeszZahE+pcAhkhlSmVGwWW0CwGofA0qq3VudXuGgGAADiKpQ3KW9eqvAOk0B1oYKcSWIbFXUCKRQcNGZWa7NUsUkctb6AEFBFyHOBAADS6B8AIgANpeFdaGzN1oAC6BrPjrvwF7K1OTzEZkPYqipyR6GtkhWMX5od50UDgGMANJNzC0EfF-J4fjRVzyEo4tGtqJDUF+HMOA9r0qoCDlMVOkHbbYDAAHKOTcAptrwK3FVrAVIkElnxS5UBMSs0+F+yJtsRyZN4Zs2jnTbAJMVB+vZksyWQs47Rx1QA). You can ignore the majority of the `parseError` implementation. Just know that it is a function that takes in an `unknown` error and returns an `ErrorResult` discriminated union. The issue is at the bottom of the reproduction in the `useGetTask` hook.

Any thoughts?

**Comment by TkDodo** [comment, marked as answer]
> The issue is at the bottom of the reproduction in the useGetTask hook.

it looks like re-assigning `result.error` doesn't change the type of `error`. I think you would need a type assertion here:

```
  if(result.isError){
    result.error = parseError(result.error);
    return result as UseQueryResult<Task, ErrorResult>
  }
```

however, it still doesn't check out because you `return result` at the very end, and there, you haven't transformed the error. Is there a reason why you can't always call `parseError` and handle the null check inside of it? That would enable you to at least do:

```
  const result = useQuery({
    queryKey: ['getTask', id],
    queryFn: () => getTask(id)
  })

  result.error = parseError(result.error);
  return result as UseQueryResult<Task, ErrorResult>
```

**Reply by Ionshard** [reply]
Thanks, always processing the error but handling `null` definitely makes sense. I really try to avoid type assertions but I can make an exception since this will be in generated code. Thank you for the feedback!

**Reply by Ionshard** [reply]
When actually implementing this I got a runtime error when attempting to set the `error` since it has only a `getter`. I currently solved this by instead doing:
```typescript
  const result = useQuery({
    queryKey: ['getTask', id],
    queryFn: () => getTask(id)
  })

  const error = parseError(result.error);
  return { ...result, error } as UseQueryResult<Task, ErrorResult>
```

Is it possible anything could go wrong with the spread notation?

**Reply by TkDodo** [reply]
it's probably due to the property tracking we do with `Object.defineProperty`:

https://github.com/TanStack/query/blob/c6748fba7bd7695b851f8c4dd457f7ac99b63d0d/packages/query-core/src/queryObserver.ts#L265-L274

The thing is, if you use `...result`, you'll invoke all getters of all fields, thus voiding the perf improvement of tracked queries. In that case, you can also set `notifyOnChangeProps: 'all'`.

More on this topic here: https://tkdodo.eu/blog/react-query-render-optimizations#tracked-queries

**Reply by TkDodo** [reply]
oh, do it in the queryFn then ?

```
  return useQuery({
    queryKey: ['getTask', id],
    queryFn: async () => { 
      try {
        return getTask(id)
      }
      catch(error) {
        throw parseError(error)
      }
    }
  }) as UseQueryResult<Task, ErrorResult>
```

**Reply by Ionshard** [reply]
I was thinking about doing it in the `queryFn` but my concern was what would happen if an error occurs in the `select` property? We're often dealing with poorly defined backend data so sometimes we will use Zod in the `select` to validate the data against a stricter schema.

EDIT: Oh ... I can just do the same pattern in the select function. I will give this a go! I didn't know about tracked queries ... but now that I do I definitely want them! 🤣

**Reply by TkDodo** [reply]
yeah it wouldn't cover errors from select. But parsing backend data is something that I would do in the queryFn. Otherwise, you'll cache the wrong response, you'll retrieve the wrong response with `queryClient.getQueryData`, and you'll need to transform with select for every subscriber.

**Reply by Ionshard** [reply]
That is valid, but we are explicitly doing it in the select because we want to avoid cache bashing other queries. So we cache the exact data that is returned from the back-end and individual uses can further narrow it down via select if they so desire without affecting any other queries. In the case of using Zod to "parse" ... no actual data transformation happens ... it's just letting Typescript know that the value adheres to a stricter schema at runtime (or we get a runtime error) but never modifies the data itself.

**Reply by TkDodo** [reply]
> no actual data transformation happens

that's fine, it's just that zod, per default, strips away unknown fields and it can also do refinements on the actual values at runtime. But yes, you don't need to do that.

**Reply by Ionshard** [reply]
Awesome, I think I got this working with the `try catch` in the `queryFn` and in the `select` function. This turns out to be far better because it also solves the problem of parsing the error before it goes to `onError` in a `useMutation` when following the same pattern.

Thank you so much for the feedback!
